Sources/SwiftProtobuf/BinaryDecoder.swift

```diff
@@ -46,12 +50,47 @@ internal struct BinaryDecoder: Decoder {
 
     private var complete: Bool {return available == 0}
 
-    internal init(forReadingFrom pointer: UnsafePointer<UInt8>, count: Int, extensions: ExtensionMap? = nil) {
+    internal init(
+      forReadingFrom pointer: UnsafePointer<UInt8>,
+      count: Int,
+      options: BinaryDecodingOptions,
+      extensions: ExtensionMap? = nil
+    ) {
         // Assuming baseAddress is not nil.
         p = pointer
         available = count
         fieldStartP = p
         self.extensions = extensions
+        self.options = options
+        recursionBudget = options.messageDepthLimit
+    }
+
+    internal init(
+      forReadingFrom pointer: UnsafePointer<UInt8>,
+      count: Int,
+      parent: BinaryDecoder
```

Comment:
Looking more closely at this (I don't remember if the JSON decoder had the same thing), since `BinaryDecoder` is a `struct`, it seems like this effectively copies the decoder when it's passed in as a parent, and then discards the copy afterwards (assuming the compiler isn't smart enough to elide the copy). Do we know if this has a negative performance impact?